File: 0763_med_hot_partition-labels/solution.py
from typing import List


class Solution:
    def partitionLabels(self, s: str) -> List[int]:

        last_occur = {}

        for i, ch in enumerate(s):
            last_occur[ch] = i

        result = []
        start = 0
        end = 0

        for i, ch in enumerate(s):
            end = max(end, last_occur[ch])
            if i == end:
                result.append(end - start + 1)
                start = i + 1

        return result

File: 0763_med_hot_partition-labels/solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    public List<Integer> partitionLabels(String s) {
        int[] hash = new int[26];

        for (int i = 0; i < s.length(); i++) {
            hash[s.charAt(i) - 'a'] = i;
        }

        List<Integer> result = new ArrayList<>();
        int start = 0;
        int end = 0;

        for (int i = 0; i < s.length(); i++) {
            end = Math.max(end, hash[s.charAt(i) - 'a']);

            if (i == end) {
                result.add(end - start + 1);
                start = i + 1;
            }
        }

        return result;
    }
}